File: android/src/main/kotlin/dev/locus/core/HeadlessValidationDispatcher.kt
package dev.locus.core

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import android.util.Log
import dev.locus.service.HeadlessValidationService
import org.json.JSONArray
import org.json.JSONObject

/**
 * Dispatcher for headless pre-sync validation.
 * 
 * When the Flutter app is terminated but location sync is needed,
 * this dispatcher invokes a registered Dart callback to validate
 * locations before syncing. This allows business logic validation
 * even when the app UI is not running.
 */
class HeadlessValidationDispatcher(
    private val context: Context,
    private val config: ConfigManager,
    private val prefs: SharedPreferences?
) {
    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Validates locations and extras via headless callback.
     * 
     * @param locations List of location data to validate
     * @param extras Additional metadata for validation
     * @param callback Callback with validation result (true = proceed, false = abort)
     * @param timeoutMs Maximum time to wait for validation response
     */
    fun validate(
        locations: List<Map<String, Any>>,
        extras: Map<String, Any>,
        callback: (Boolean) -> Unit,
        timeoutMs: Long = 10_000L
    ) {
        if (!config.enableHeadless || prefs == null) {
            Log.d(TAG, "Headless disabled or no prefs, allowing sync")
            callback(true)
            return
        }

        val dispatcher = prefs.getLong(KEY_VALIDATION_DISPATCHER, 0L)
        val validationCallback = prefs.getLong(KEY_VALIDATION_CALLBACK, 0L)

        // If no validation callback registered, allow sync by default
        if (dispatcher == 0L || validationCallback == 0L) {
            Log.d(TAG, "No validation callback registered, allowing sync")
            callback(true)
            return
        }

        runCatching {
            val locationsJson = JSONArray().apply {
                locations.forEach { loc ->
                    put(JSONObject(loc))
                }
            }

            val extrasJson = JSONObject(extras)

            val payload = JSONObject().apply {
                put("type", "validatePreSync")
                put("locations", locationsJson)
                put("extras", extrasJson)
            }

            val intent = Intent(context, HeadlessValidationService::class.java).apply {
                putExtra("dispatcher", dispatcher)
                putExtra("callback", validationCallback)
                putExtra("payload", payload.toString())
                putExtra("timeoutMs", timeoutMs)
            }

            HeadlessValidationService.enqueueWork(context, intent) { result ->
                mainHandler.post { callback(result) }
            }
        }.onFailure { e ->
            Log.e(TAG, "Failed to dispatch validation: ${e.message}")
            // Proceed with sync on error to avoid blocking permanently
            callback(true)
        }
    }

    /**
     * Checks if headless validation is available.
     */
    fun isAvailable(): Boolean {
        if (!config.enableHeadless || prefs == null) return false
        val dispatcher = prefs.getLong(KEY_VALIDATION_DISPATCHER, 0L)
        val validationCallback = prefs.getLong(KEY_VALIDATION_CALLBACK, 0L)
        return dispatcher != 0L && validationCallback != 0L
    }

    companion object {
        private const val TAG = "locus.HeadlessValidation"
        const val KEY_VALIDATION_DISPATCHER = "bg_headless_validation_dispatcher"
        const val KEY_VALIDATION_CALLBACK = "bg_headless_validation_callback"
    }
}

File: ios/Classes/Core/HeadlessValidationDispatcher.swift
import Flutter
import Foundation

/// Dispatcher for headless pre-sync validation on iOS.
///
/// When the Flutter app is terminated but location sync is needed,
/// this dispatcher invokes a registered Dart callback to validate
/// locations before syncing. This allows business logic validation
/// even when the app UI is not running.
class HeadlessValidationDispatcher {
    private let config: ConfigManager
    private var validationEngine: FlutterEngine?
    private let engineName = "locus_validation_engine"
    private let channelName = "locus/headless_validation"
    
    private var pendingValidation: ((Bool) -> Void)?
    private var validationTimeout: DispatchWorkItem?
    
    init(config: ConfigManager) {
        self.config = config
    }
    
    deinit {
        cleanup()
    }
    
    /// Checks if headless validation is available.
    var isAvailable: Bool {
        guard config.enableHeadless else { return false }
        let dispatcher = SecureStorage.shared.getInt64(forKey: SecureStorage.validationDispatcherKey) ?? 0
        let callback = SecureStorage.shared.getInt64(forKey: SecureStorage.validationCallbackKey) ?? 0
        return dispatcher != 0 && callback != 0
    }
    
    /// Validates locations and extras via headless callback.
    ///
    /// - Parameters:
    ///   - locations: List of location data to validate
    ///   - extras: Additional metadata for validation
    ///   - timeoutSeconds: Maximum time to wait for validation response
    ///   - completion: Callback with validation result (true = proceed, false = abort)
    func validate(
        locations: [[String: Any]],
        extras: [String: Any],
        timeoutSeconds: TimeInterval = 10,
        completion: @escaping (Bool) -> Void
    ) {
        guard config.enableHeadless else {
            log("Headless disabled, allowing sync")
            completion(true)
            return
        }
        
        guard let dispatcher = SecureStorage.shared.getInt64(forKey: SecureStorage.validationDispatcherKey),
              let callback = SecureStorage.shared.getInt64(forKey: SecureStorage.validationCallbackKey),
              dispatcher != 0, callback != 0 else {
            log("No validation callback registered, allowing sync")
            completion(true)
            return
        }
        
        // Only one validation at a time
        if pendingValidation != nil {
            log("Validation already in progress, allowing sync")
            completion(true)
            return
        }
        
        pendingValidation = completion
        
        // Set up timeout
        let timeout = DispatchWorkItem { [weak self] in
            self?.log("Validation timed out, allowing sync")
            self?.completeValidation(result: true)
        }
        validationTimeout = timeout
        DispatchQueue.main.asyncAfter(deadline: .now() + timeoutSeconds, execute: timeout)
        
        // Start headless engine if needed
        DispatchQueue.main.async { [weak self] in
            self?.executeValidation(dispatcher: dispatcher, callback: callback, locations: locations, extras: extras)
        }
    }
    
    // MARK: - Private Methods
    
    private func executeValidation(
        dispatcher: Int64,
        callback: Int64,
        locations: [[String: Any]],
        extras: [String: Any]
    ) {
        // Create engine if needed
        if validationEngine == nil {
            guard let callbackInfo = FlutterCallbackCache.lookupCallbackInformation(dispatcher) else {
                log("Could not lookup callback info for dispatcher")
                completeValidation(result: true)
                return
            }
            
            let engine = FlutterEngine(name: engineName, project: nil, allowHeadlessExecution: true)
            guard engine.run(withEntrypoint: callbackInfo.callbackName, libraryURI: callbackInfo.callbackLibraryPath) else {
                log("Failed to start headless engine")
                completeValidation(result: true)
                return
            }
            validationEngine = engine
        }
        
        guard let engine = validationEngine else {
            log("No engine available")
            completeValidation(result: true)
            return
        }
        
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: engine.binaryMessenger)
        
        let payload: [String: Any] = [
            "type": "validatePreSync",
            "locations": locations,
            "extras": extras
        ]
        
        let args: [String: Any] = [
            "callbackHandle": callback,
            "payload": payload
        ]
        
        channel.invokeMethod("validatePreSync", arguments: args) { [weak self] result in
            let validated = result as? Bool ?? true
            self?.log("Validation result: \(validated)")
            self?.completeValidation(result: validated)
        }
        
        // Schedule engine cleanup
        DispatchQueue.main.asyncAfter(deadline: .now() + 30) { [weak self] in
            self?.cleanupIfIdle()
        }
    }
    
    private func completeValidation(result: Bool) {
        validationTimeout?.cancel()
        validationTimeout = nil
        
        let completion = pendingValidation
        pendingValidation = nil
        completion?(result)
    }
    
    private func cleanupIfIdle() {
        guard pendingValidation == nil else { return }
        cleanup()
    }
    
    private func cleanup() {
        validationTimeout?.cancel()
        validationTimeout = nil
        pendingValidation = nil
        validationEngine?.destroyContext()
        validationEngine = nil
    }
    
    private func log(_ message: String) {
        #if DEBUG
        print("[HeadlessValidation] \(message)")
        #endif
    }
    
    // MARK: - Static Registration Methods
    
    /// Registers the headless validation callback handles.
    /// - Parameters:
    ///   - dispatcher: The dispatcher callback handle
    ///   - callback: The validation callback handle
    static func registerCallback(dispatcher: Int64, callback: Int64) {
        _ = SecureStorage.shared.setInt64(dispatcher, forKey: SecureStorage.validationDispatcherKey)
        _ = SecureStorage.shared.setInt64(callback, forKey: SecureStorage.validationCallbackKey)
    }
}
